src/analyses/goto_check.cpp

```diff
@@ -1226,21 +1230,21 @@ void goto_checkt::pointer_primitive_check(
                        : size_of_expr_opt.value();
 
   const conditionst &conditions = address_check(pointer, size);
-
-  exprt::operandst conjuncts;
-
   for(const auto &c : conditions)
-    conjuncts.push_back(c.assertion);
-
-  const or_exprt or_expr(null_object(pointer), conjunction(conjuncts));
-
-  add_guarded_property(
-    or_expr,
-    "pointer in pointer primitive is neither null nor valid",
-    "pointer primitives",
-    expr.source_location(),
-    expr,
-    guard);
+  {
+    // we want all address checks *except* for the null check because these
+    // should only fail if the assumption is on in invalid non-null pointer.
+    if(c.description != pointer_is_null_check_description)
```

Comment:
This seems to have less checks?  Thanks for adding the comment but I think I am struggling to understand it, possibly because I am missing the context.  Please can you explain it a bit more to me?

Reply:
Adding to this: is a string comparison really the best solution? Couldn't we instead compare `c.assertion` to `null_object(pointer)`?

Reply:
@tautschnig I don’t like it either, but I don’t like repeating the implementation of `c.assertion` much better. My preferred thing to do would be to add stable IDs to the checks we generate so we can check that, e.g. this is "the address null check", something along those lines.

Reply:
How about adding further members to `conditiont` that track which kinds of properties are already asserted? We don't have to be complete right away, you could just start with a single one: `bool has_null_check`.

Reply:
@tautschnig One option to get this through right now would be to park the idea of how to filter out checks for later and just accept the redundant null-check for now, but that doesn’t make me happy either.

Reply:
Well, why not try to route of adding, e.g., `has_null_check`?

Reply:
@tautschnig I don’t completely understand how you’d want that to work – I want to remove the null check, not ask whether it has one.

Reply:
I'd basically use a `bool` member instead of the string. But then what @thomasspriggs [said](https://github.com/diffblue/cbmc/pull/5582/files#r525275170) seems a lot better.